crates/iota-core/src/authority.rs

```diff
@@ -4705,6 +4688,65 @@ impl AuthorityState {
         (next_protocol_version, system_packages)
     }
 
+    /// Returns the indices of validators that support the given protocol
+    /// version. This includes both committee and non-committee validators
+    /// based on their capabilities. Uses active validators instead of committee
+    /// indices.
+    fn get_validators_supporting_protocol_version(
+        target_protocol_version: ProtocolVersion,
+        active_validators: Vec<(AuthorityName, AuthorityPublicKey)>,
+        capabilities: Vec<AuthorityCapabilitiesV1>,
+    ) -> Vec<u64> {
+        let mut eligible_validators = Vec::new();
+
+        for capability in capabilities {
+            // Check if this validator supports the target protocol version and digest
+            if capability
+                .supported_protocol_versions
+                .get_version_digest(target_protocol_version)
+                .is_some()
+            {
+                // Find the validator's index in the active validators list
+                if let Some(index) = active_validators
+                    .iter()
+                    .position(|(name, _)| *name == capability.authority)
+                {
+                    eligible_validators.push(index as u64);
+                }
+            }
+        }
+
+        // Sort indices for deterministic behavior
+        eligible_validators.sort();
+        eligible_validators
+    }
+
+    /// Calculates the sum of weights for eligible validators that are part of
+    /// the committee. Takes the indices from
+    /// get_validators_supporting_protocol_version and maps them back
+    /// to committee members to get their weights.
+    fn calculate_eligible_validators_weight(
+        eligible_validator_indices: &[u64],
+        active_validators: &[(AuthorityName, AuthorityPublicKey)],
+        committee: &Committee,
+    ) -> u64 {
+        let mut total_weight = 0u64;
+
+        for &index in eligible_validator_indices {
+            if let Some((authority_name, _)) = active_validators.get(index as usize) {
```

Comment:
Why not just use index operator? The indices must be consistent with the list of validators:
```suggestion
            let (authority_name, _) = active_validators[index as usize];
```

Reply:
done